docs: arbchain-custom-genesis-state (#3206)
OffchainLabs/arbitrum-docs: docs/launch-arbitrum-chain/02-configure-your-chain/common/validation-and-security/custom-genesis-state.mdx

Initializing from a genesis state would be helpful in the following scenarios:

1. Redeploying a new testnet with the pre-existing state if the current testnet is broken.
2. Deploying additional sibling chains with pre-existing contracts.
3. Simplifying self-serve backends for RaaSes, such that hundreds of testnets can be deployed with similar contracts.
4. Reducing lift on third-party infrastructure teams to redeploy contracts for new chains.

## Why should I use a custom genesis state?

Using a custom genesis allows you to initialize a new blockchain with a customized genesis state and network configuration through the following options:

- **Predeployed contracts**: standard chains start without any smart contracts. This feature allows you to preload contract bytecode in the very first block, so infrastructure is available at launch.
- **Initial account state (allocations)**: you can pre-configure the ledger—including account balances and contract storage—before the network opens for transactions.

> - **Enable advanced features**: chain operators can use this feature to launch a chain with advanced customizations, such as [minting/burning gas tokens via third-party bridges](http:///launch-arbitrum-chain/features/common/gas-and-fees/choose-native-mint-burn) and compliance-focused transaction filtering.

Reviewer: ```suggestion
- **Enable advanced features**: chain operators can use this feature to launch a chain with advanced customizations, such as [minting/burning gas tokens via third-party bridges](http://launch-arbitrum-chain/features/common/gas-and-fees/choose-native-mint-burn) and compliance-focused transaction filtering.
```

| `L1_BASE_FEE`      | The initial L1 gas price (in wei) used to calibrate the chain initiation. |
| `NITRO_NODE_IMAGE` | The Nitro node Docker image used for hashing and node operations.         |

#### Deployment configuration (optional)

This step is only required if you choose to deploy the rollup to the parent chain (Step 3 below).

| Variable                 | Description                                                            |
| ------------------------ | ---------------------------------------------------------------------- |
| DEPLOYER_PRIVATE_KEY     | Private key of the account responsible for the rollup deployment.      |
| BATCH_POSTER_PRIVATE_KEY | Private key for the sequencer's batch-posting address.                 |
| VALIDATOR_PRIVATE_KEY    | Private key for the validator/bonder address                           |
| PARENT_CHAIN_RPC         | RPC endpoint for the parent chain (e.g., Arbitrum Sepolia or Ethereum) |

#### Execution steps

1. Prepare the environment

   - From the repository root, install dependencies and navigate to the generator directory.

   ```shell
   yarn install && yarn build

>    cd examples/generate-genesis-file && cp .env/example .env

Reviewer: ```suggestion
   cd examples/generate-genesis-file && cp .env.example .env
```

   ```

2. Generate gensis

   - Ensure your .env is configured and run the dev script.

   ```shell

>    bash yarn dev

Reviewer: ```suggestion
   ```shell
   yarn dev
```

3. Simplifying self-serve backends for RaaSes, such that hundreds of testnets can be deployed with similar contracts.
4. Reducing lift on third-party infrastructure teams to redeploy contracts for new chains.

## Why should I use a custom genesis state?

Using a custom genesis allows you to initialize a new blockchain with a customized genesis state and network configuration through the following options:

- **Predeployed contracts**: standard chains start without any smart contracts. This feature allows you to preload contract bytecode in the very first block, so infrastructure is available at launch.
- **Initial account state (allocations)**: you can pre-configure the ledger—including account balances and contract storage—before the network opens for transactions.
- **Enable advanced features**: chain operators can use this feature to launch a chain with advanced customizations, such as [minting/burning gas tokens via third-party bridges](/launch-arbitrum-chain/features/common/gas-and-fees/choose-native-mint-burn.mdx) and compliance-focused transaction filtering.

<VanillaAdmonition type="danger" title="Preallocated balances are not backed by the parent chain">

> Preallocating account balances in the genesis state credits ETH (or your custom gas token) on your Arbitrum chain **without** depositing anything into the parent-chain bridge. The protocol does **not** enforce that the bridge holds ETH equal to the balances you preallocate, and there is no deploy-time check.

Author: ```suggestion
Preallocating account balances in the genesis state credits **ETH** (or your custom gas token) on your Arbitrum chain **without** depositing anything into the parent-chain bridge. The protocol does **not** require the bridge to hold ETH equal to the balances you preallocate, and there is no deploy-time check.
```
